clang/lib/CodeGen/ModuleBuilder.cpp

```diff
@@ -318,6 +318,9 @@ namespace {
       if (Diags.hasUnrecoverableErrorOccurred())
         return;
 
+      if (RD->shouldEmitInExternalSource())
```

Comment:
Maybe this check should be part of DefineUsedVTables()?

Reply:
I use `isInCurrentModuleUnit` instead since I feel the semantics are more clear.

Reply:
Did you mean to change something in ModuleBuilder.cpp?

Reply:
Yes, I think it is not bad to put the check `RD->shouldEmitInExternalSource()` into `HandleVTable`. I feel the most natural logic may be, I see a vtable then I pass it to the consumer and it is the job of the consumer to decide whether or not to emit it. WDYT?

Reply:
The way I see it, Sema should have the exact right answer for whether the vtable is required.  If we need to second-guess the choice later, that means Sema didn't compute the right thing in the first place, which would be a bug in Sema.

Reply:
Done. I don't mind doing it really. But what I confuse is, it looks there are a lot of codes in CodeGen to decide whether or not to emit the vtable. So it looks not like a pure Sema job to me.

Reply:
Most of the CodeGen code surrounding vtable emission is optimizations.  Specifically, to avoid emitting vtables which are odr-used but not actually necessary to emit, and to allow emitting available_externally vtables.  If we didn't care about those optimizations, the code would be pretty simple.

Reply:
Thanks. It makes sense.